crates/dojo-lang/src/contract.rs

```diff
@@ -197,8 +200,232 @@ impl DojoContract {
             .to_string(),
         )]
     }
-}
 
-// fn is_context(db: &dyn SyntaxGroup, param: &Param) -> bool {
-//     param.type_clause(db).ty(db).as_syntax_node().get_text(db) == "Context"
-// }
+    /// Gets name, modifiers and type from a function parameter.
+    pub fn get_parameter_info(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        param: ast::Param,
+    ) -> (String, String, String) {
+        let name = param.name(db).text(db).trim().to_string();
+        let modifiers = param.modifiers(db).as_syntax_node().get_text(db).trim().to_string();
+        let param_type =
+            param.type_clause(db).ty(db).as_syntax_node().get_text(db).trim().to_string();
+
+        (name, modifiers, param_type)
+    }
+
+    /// Check if the function has a self parameter.
+    ///
+    /// Returns
+    ///  * a boolean indicating if `self` has to be added,
+    //   * a boolean indicating if there is a `ref self` parameter.
+    pub fn check_self_parameter(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        param_list: ast::ParamList,
+    ) -> (bool, bool) {
+        let mut add_self = true;
+        let mut has_ref_self = false;
+        if !param_list.elements(db).is_empty() {
+            let (param_name, param_modifiers, param_type) =
+                self.get_parameter_info(db, param_list.elements(db)[0].clone());
+
+            if param_name.eq(&"self".to_string()) {
+                if param_modifiers.contains(&"ref".to_string())
+                    && param_type.eq(&"ContractState".to_string())
+                {
+                    has_ref_self = false;
+                    add_self = false;
+                }
+
+                if param_type.eq(&"@ContractState".to_string()) {
+                    add_self = false;
+                }
+            }
+        };
+
+        (add_self, has_ref_self)
+    }
+
+    /// Check if the function has multiple IWorldDispatcher parameters.
+    ///
+    /// Returns
+    ///  * a boolean indicating if the function has multiple world dispatchers.
+    pub fn check_world_dispatcher(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        param_list: ast::ParamList,
+    ) -> bool {
+        let mut count = 0;
+
+        param_list.elements(db).iter().for_each(|param| {
+            let (_, _, param_type) = self.get_parameter_info(db, param.clone());
```

Comment:
We should check that the `param_name` **is** `world` as it is expanded as such and we only authorize one `IWorldDispatcher`:
```rust
"let world = self.world_dispatcher.read();\n"
```

Reply:
In fact, the `check_world_dispatcher` function counts the number of parameters of type `IWorldDispatcher` and indicates if there is more than one.

Then, in `rewrite_parameters`, the `world: IWorldDispatcher` parameter is removed only if:
- there is only one `IWorldDispatcher` (which is `world`),
- `world` is the first parameter excluding `self` (so, if `self` is not there, index = 0, if `self` is already there, index=1).

Finally, if `world` has been removed, the statement to read `world` from `world_dispatcher` is added.

Maybe I should rename the `check_world_dispatcher` to something like `check_multiple_world_dispatchers`.

Is it fine for you now or did I miss something ? 🤔

Reply:
Yep, totally agree on the flow and it's great.

The only caveat is that if I use `ahah_world: IWorldDispatcher` instead of `world: IWorldDispatcher`, the compilation fails with the same cryptic error of wrong number of arguments. Because in the `rewrite_parameters` the `else` is not an error, just a re-write.

The naming is good to me. :+1:

Reply:
Ah yes, I see !
So the rule is: if there is a `IWorldDispatcher` parameter, it must be:
- the first parameter,
- named `world`,
- without any other `IWorldDispatcher` parameters.
👍

Reply:
> Ah yes, I see !
> 
> So the rule is: if there is a `IWorldDispatcher` parameter, it must be:
> 
> - the first parameter,
> 
> - named `world`,
> 
> - without any other `IWorldDispatcher` parameters.
> 
> 👍 

Nicely resumed, yes! Sorry for lack of precision on the first review round.

If one do want an additional world to interact with, they can always pass an address (in the interface too then), and build the dispatcher into the system. 👍